fix(go/adbc/driver/snowflake): Removing SQL injection to get table name with special character for getObjectsTables

AnithaPanduranganMS opened the pull request:
**Description:**

> GetObjects API was inconsistent getting table with special character and making the conditions case-insensitive.

**Solution:**

> Passing table names as query argument and avoiding SQL Injection

**Testing:**

> Added test in DriverTest

Fixes #1225

zeroshade reviewed (commented):
Can we add a test to confirm that we've removed the potential for sql injection here?

jduo commented on csharp/test/Drivers/Snowflake/DriverTests.cs:
```diff
@@ -39,6 +39,8 @@ public class DriverTests : IDisposable
         readonly AdbcDriver _snowflakeDriver;
         readonly AdbcDatabase _database;
         readonly AdbcConnection _connection;
+        readonly AdbcStatement _statement;
```
This needs to be Dispose()d too.

jduo commented on go/adbc/driver/snowflake/connection.go:
```diff
@@ -281,11 +281,14 @@ func (c *cnxn) getObjectsDbSchemas(ctx context.Context, depth adbc.ObjectDepth,
+	queryArgs := make([]interface{}, 0, 2)
 	if catalog != nil && *catalog != "" {
-		conditions = append(conditions, ` CATALOG_NAME LIKE '`+*catalog+`'`)
+		conditions = append(conditions, ` CATALOG_NAME ILIKE ? `)
```
The PR comments should indicate that you're also making the conditions case-insensitive.

AnithaPanduranganMS commented:
> Can we add a test to confirm that we've removed the potential for sql injection here?

Added test in C# driver test

ryan-syed commented on go/adbc/driver/snowflake/connection.go:
```diff
@@ -281,11 +281,14 @@ func (c *cnxn) getObjectsDbSchemas(ctx context.Context, depth adbc.ObjectDepth,
 	}
 
 	conditions := make([]string, 0)
+	queryArgs := make([]interface{}, 0, 2)
```
Is there a specific reason to use `interface` here, when the type is know to be `string`?

AnithaPanduranganMS replied:
Query arguments only accepts interface

ryan-syed commented on go/adbc/driver/snowflake/connection.go:
```diff
@@ -828,7 +836,7 @@ func (c *cnxn) GetTableSchema(ctx context.Context, catalog *string, dbSchema *st
 	if dbSchema != nil {
 		tblParts = append(tblParts, strconv.Quote(strings.ToUpper(*dbSchema)))
 	}
-	tblParts = append(tblParts, strconv.Quote(strings.ToUpper(tableName)))
```
Should `string.ToUpper` conversion be avoided for catalog and schema name too?

ryan-syed commented on csharp/test/Drivers/Snowflake/DriverTests.cs:
```diff
@@ -279,17 +279,51 @@ public void CanGetObjectsAll()
+
+            List<AdbcTable> tables = catalogs
+                .Where(s => s.Name.Equals(databaseName))
+                .Select(s => s.DbSchemas)
```
We probably need to add `WHERE` filtering for `schemaName` and `tableName` too as they can contain wild card characters and using only `FirstOrDefault` could end up with a potential different indentifier.

ryan-syed reviewed (changes requested):
Consider adding tests for special characters in catalog and schema names too. Also, temporary tables that have a wide range of parametrized tests should help improve reliability.

AnithaPanduranganMS commented:
Implemented on top of https://github.com/apache/arrow-adbc/pull/1352 (perf(go/adbc/driver/snowflake): Reduced SQL calls in GetObjects to two, added prefixing DbName for INFORMATION_SCHEMA)

lidavidm commented:
@AnithaPanduranganMS do you mind rebasing?
